src/gfn/examples/box_utils.py

```diff
@@ -0,0 +1,457 @@
+        logprobs[torch.norm(1 - self.centers, dim=1) <= self.delta] = 0.0
+        return logprobs
+
+
+class BoxPFEStimator(ProbabilityEstimator):
+    r"""Estimator for P_F for the Box environment. Uses the BoxForwardDist distribution"""
+
+    def __init__(
+        self,
+        env: BoxEnv,
+        module: torch.nn.Module,
+        n_components_s0: int,
+        n_components: int,
+    ):
+        super().__init__(env, module)
+        self.n_components_s0 = n_components_s0
+        self.n_components = n_components
+
+    def check_output_dim(
+        self, module_output: TT["batch_shape", "output_dim", float]
+    ) -> None:
+        # Not implemented because the module output shape is different for s_0 and for other states
+        pass
+
+    def to_probability_distribution(
+        self, states: States, module_output: TT["batch_shape", "output_dim", float]
+    ) -> Distribution:
+        # First, we verify that the batch shape of states is 1
+        assert len(states.batch_shape) == 1
+        # Then, we check that if one of the states is [0, 0] then all of them are
+        # TODO: is there a way to bypass this ? Could we write a custom distribution
+        # TODO: that sometimes returns a QuarterDisk and sometimes a QuarterCircle(northwestern=True) ?
+        if torch.any(states == 0.0):
+            assert torch.all(states == 0)
+            # we also check that module_output is of shape n_components_s0 * 5
+            assert module_output.shape == (self.n_components_s0, 5)
+            # In this case, we use the QuarterDisk distribution
+            mixture_logits, alpha_r, beta_r, alpha_theta, beta_theta = torch.split(
+                module_output, 1, dim=-1
+            )
+            return QuarterDisk(
+                delta=self.env.delta,
+                mixture_logits=mixture_logits,
+                alpha_r=alpha_r,
+                beta_r=beta_r,
+                alpha_theta=alpha_theta,
+                beta_theta=beta_theta,
+            )
+        else:
+            # we check that the module_output is of shape (*batch_shape, 1 + 3 * n_components)
+            assert module_output.shape == states.batch_shape + (
+                1 + 3 * self.n_components,
+            )
+            # In this case, we use the QuarterCircleWithExit distribution
+            exit_probability, mixture_logits, alpha, beta = torch.split(
+                module_output,
+                [1, self.n_components, self.n_components, self.n_components],
+                dim=-1,
+            )
+            exit_probability = exit_probability.squeeze(-1)
+            return QuarterCircleWithExit(
+                delta=self.env.delta,
+                northeastern=True,
+                centers=states,
+                exit_probability=exit_probability,
+                mixture_logits=mixture_logits,
+                alpha=alpha,
+                beta=beta,
+            )
+
+
+class BoxPBEstimator(ProbabilityEstimator):
```

Comment:
I'm skeptical about the requirement to define an environment specific estimator. I feel like this shouldn't be required if we had the correct abstraction. I wonder what @vict0rsch thinks?

Reply:
Let me summarize the expected workflow with the current codebase, and then please let me know what abstractions we can change to make it better:
Given an environment, how do we sample trajectories ?
- We need a distribution over actions that is conditioned on the state. For discrete environments for example, the distribution is a Categorical one. It can be anything we want for other environments. So there is always **a need for a `torch.Distribution`**.
- - Example: For the Box environment, in this PR, I coded Mixture of Betas (scaled). Why ? The environment is bounded, so we need a parametrized distribution that gives us actions within a bound. Beta's sample space is [0, 1], but is usually unimodal. Thus the mixture, and the scaling.
- This distribution, whether it's categorical or Beta, or whatever, has some parameters $\theta(s)$. For example, for Beta, it's the concentration parameters $\alpha$ and $\beta$.
- $\theta(s)$ is usually a neural network. So there is, almost always, **a need for a `nn.Module`, that takes as input a batch of states, and returns the parameters of the distribution**
- The `PFEstimator` and `PBEstimator` serve just one purpose: **transforming the output of the `nn.Module`, to the `Distribution`**, eventually by applying some modifications (please see the new PR for an updated version of this file, where there are some modifications - the scaling -, and there is a Neural Net implemented).

An alternative that is aligned with your suggestion (i.e. not asking the user to code all 3 of: `Distribution`, `nn.Module`, `PF/BEstimator`), is to do the postprocessing of `nn.Module`'s output, in the `nn.Module`. So we would need some abstract `GFNModule` class that subclasses `nn.Module` but requires the implementation of `self.to_probaiblity_distribution(...)` that is now in the `ProbabilityEstimator` class.

Thoughts ?